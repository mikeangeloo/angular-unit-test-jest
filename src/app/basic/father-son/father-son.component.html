<h5>Hijo</h5>
<code>{{client | json}}</code>
<br><br>
<button *ngIf="client" class="btn btn-outline-primary" (click)="onChange(5)">Cambiar Id</button>
<button *ngIf="client" class="btn btn-outline-danger" (click)="onDelete()" data-test="btn-delete">Borrar Cliente</button>
