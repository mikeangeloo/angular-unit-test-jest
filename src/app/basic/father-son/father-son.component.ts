import {Component, EventEmitter, Input, OnInit, Output} from '@angular/core';
import {ClientI} from '../interfaces/client.interface';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.scss']
})
export class FatherSonComponent {

  @Input() client?: ClientI;
  @Output() onDeleteClient = new EventEmitter();
  @Output() onClientUpdated = new EventEmitter<ClientI>();

  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }

  onChange(id: number) {
    if (!this.client) return;
    //this.client.id = newId; // no usar
    this.client = {... this.client, id} // usamos desustruraci√≥n
    this.onClientUpdated.emit({... this.client})
  }
}
