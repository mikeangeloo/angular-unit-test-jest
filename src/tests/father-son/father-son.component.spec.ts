import { ComponentFixture, TestBed } from '@angular/core/testing';
import {FatherSonComponent} from '../../app/basic/father-son/father-son.component';


describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ FatherSonComponent ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('debe ser hacer match on el snapshopt', () => {
    expect(compiled).toMatchSnapshot();
  })

  it('no deben aparecer botones si no hay cliente', () => {
    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(0)
  })

  it('deben aparecer 2 botones si  hay cliente', () => {
    component.client = {id: 1, name: 'Juan'}
    fixture.detectChanges();
    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(2)
  })


  it('si hay cliente, hacer match con e snapshot', () => {
    component.client = {id: 1, name: 'Juan'}
    fixture.detectChanges();
    expect(compiled).toMatchSnapshot()
  })

  it ('debe de emitir onDeleteClient con el boton de eliminar', () => {
    component.client = {id: 1, name: 'Juan'}
    fixture.detectChanges();
    jest.spyOn(component.onDeleteClient, 'emit');
    const btnDelete = compiled.querySelector('[data-test=btn-delete]');
    btnDelete?.dispatchEvent(new Event('click'));
    expect(component.onDeleteClient.emit).toHaveBeenCalled();
    console.log(btnDelete?.innerHTML);
  })

});
