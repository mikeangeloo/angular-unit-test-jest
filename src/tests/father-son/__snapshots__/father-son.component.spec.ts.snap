// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent debe ser hacer match on el snapshopt 1`] = `
<div
  id="root1"
>
  <h5>
    Hijo
  </h5>
  <code>
    
  </code>
  <br />
  <br />
  
  
</div>
`;

exports[`FatherSonComponent si hay cliente, hacer match con e snapshot 1`] = `
<div
  id="root4"
>
  <h5>
    Hijo
  </h5>
  <code>
    {
  "id": 1,
  "name": "Juan"
}
  </code>
  <br />
  <br />
  <button
    class="btn btn-outline-primary"
  >
    Cambiar Id
  </button>
  
  <button
    class="btn btn-outline-danger"
    data-test="btn-delete"
  >
    Borrar Cliente
  </button>
  
</div>
`;
